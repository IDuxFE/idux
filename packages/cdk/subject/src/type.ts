export interface Observer<T> {
  (value: T): void
}

export interface Subscription {
  unsubscribe: () => void
}
