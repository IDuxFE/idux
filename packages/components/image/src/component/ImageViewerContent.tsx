/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ImageViewerContentProps } from '../types'
import type { ImageViewerConfig } from '@idux/components/config'
import type { ComputedRef, Ref } from 'vue'

import { computed, defineComponent, normalizeClass, onBeforeUnmount, onMounted, ref, watch, watchEffect } from 'vue'

import { debounce } from 'lodash-es'

import { isFirefox } from '@idux/cdk/platform'
import { useControlledProp } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

import { imageViewerContentProps } from '../types'

const mousewheelEventName = isFirefox ? 'DOMMouseScroll' : 'mousewheel'
const debounceTime = 10

type ScaleType = 'in' | 'out'
type RotateType = 'left' | 'right'
type GoType = 'previous' | 'next'

interface OprType {
  icon: string
  key: string
  opr: () => void
  visible: boolean
  disabled?: boolean
}

export default defineComponent({
  name: 'IxImageViewerContent',
  props: imageViewerContentProps,
  setup(props) {
    const config = useGlobalConfig('imageViewer')
    const zoom = useZoomRange(props, config)
    const maskClosable = useMaskClosable(props, config)
    const [visible, setVisible] = useControlledProp(props, 'visible', false)
    const { calcTransform, scaleDisabled, rotateHandle, scaleHandle, resetTransform } = useStyleOpr(zoom)
    const { activeIndex, switchDisabled, switchVisible, goHandle } = useSwitch(props, config)
    const oprList = useOprList(
      {
        goNext: () => goHandle('next'),
        goPrevious: () => goHandle('previous'),
        rotateLeft: () => rotateHandle('left'),
        rotateRight: () => rotateHandle('right'),
        zoomOut: () => scaleHandle('out'),
        zoomIn: () => scaleHandle('in'),
        close: () => setVisible(false),
      },
      scaleDisabled,
      switchDisabled,
      switchVisible,
    )
    const { onWheelScroll, onKeydown } = getImageEvent(visible, { setVisible, scaleHandle, goHandle })
    const onClickLayer = () => maskClosable.value && setVisible(false)

    onMounted(() => {
      window.addEventListener(mousewheelEventName, onWheelScroll, { passive: false, capture: false })
      window.addEventListener('keydown', onKeydown, false)
    })

    onBeforeUnmount(() => {
      window.removeEventListener(mousewheelEventName, onWheelScroll)
      window.removeEventListener('keydown', onKeydown)
    })

    watch([visible, activeIndex], ([visible$$]) => {
      visible$$ && resetTransform()
    })

    return () => (
      <div class={props.mergedPrefixCls}>
        {renderOprNode(props, oprList)}
        {renderPreviewImg(props, calcTransform, activeIndex, onClickLayer)}
      </div>
    )
  },
})

function renderOprNode(props: ImageViewerContentProps, oprList: ComputedRef<OprType[]>) {
  return (
    <div class={`${props.mergedPrefixCls}-opr`}>
      {oprList.value
        .filter(item => item.visible)
        .map(item => {
          const iconClasses = computed(() =>
            normalizeClass([
              `${props.mergedPrefixCls}-opr-item`,
              { [`${props.mergedPrefixCls}-opr-item-disabled`]: item.disabled },
            ]),
          )
          return <IxIcon class={iconClasses.value} name={item.icon} onClick={item.opr} key={item.key}></IxIcon>
        })}
    </div>
  )
}

function renderPreviewImg(
  props: ImageViewerContentProps,
  calcTransform: ComputedRef<Record<string, string>>,
  activeIndex: ComputedRef<number>,
  onClickLayer: () => void,
) {
  const curImgSrc = (props.images ?? [])[activeIndex.value]
  return (
    <div class={`${props.mergedPrefixCls}-preview`} onClick={onClickLayer}>
      <img class={`${props.mergedPrefixCls}-preview-img`} src={curImgSrc} style={calcTransform.value} />
    </div>
  )
}

function useOprList(
  { goNext, goPrevious, rotateLeft, rotateRight, zoomOut, zoomIn, close }: Record<string, () => void>,
  scaleDisabled: ComputedRef<Record<ScaleType, boolean>>,
  switchDisabled: ComputedRef<Record<GoType, boolean>>,
  switchVisible: ComputedRef<boolean>,
): ComputedRef<OprType[]> {
  return computed(() => [
    {
      key: 'goPrevious',
      icon: 'left',
      opr: goPrevious,
      disabled: switchDisabled.value.previous,
      visible: switchVisible.value,
    },
    { key: 'goNext', icon: 'right', opr: goNext, disabled: switchDisabled.value.next, visible: switchVisible.value },
    { key: 'rotateLeft', icon: 'rotate-left', opr: rotateLeft, visible: true },
    { key: 'rotateRight', icon: 'rotate-right', opr: rotateRight, visible: true },
    { key: 'zoomOut', icon: 'zoom-out', opr: zoomOut, disabled: scaleDisabled.value.out, visible: true },
    { key: 'zoomIn', icon: 'zoom-in', opr: zoomIn, disabled: scaleDisabled.value.in, visible: true },
    { key: 'close', icon: 'close', opr: close, visible: true },
  ])
}

function useSwitch(props: ImageViewerContentProps, config: ImageViewerConfig) {
  const [activeIndex, setIndex] = useControlledProp(props, 'activeIndex', 0)
  const loop = computed(() => props.loop ?? config.loop)
  const switchDisabled = computed(() => ({
    previous: !loop.value && activeIndex.value === 0,
    next: !loop.value && activeIndex.value === props.images.length - 1,
  }))
  const switchVisible = computed(() => props.images.length > 1)
  const goHandle = debounce((direction: GoType = 'next') => {
    if (direction === 'next') {
      if (switchDisabled.value.next) {
        return
      }
      setIndex(activeIndex.value >= props.images.length - 1 ? 0 : activeIndex.value + 1)
      return
    }
    if (switchDisabled.value.previous) {
      return
    }
    setIndex(activeIndex.value <= 0 ? props.images.length - 1 : activeIndex.value - 1)
  }, debounceTime)

  return {
    activeIndex,
    switchDisabled,
    switchVisible,
    goHandle,
  }
}

function useStyleOpr(zoom: ComputedRef<number[]>) {
  const initScale = computed(() => getInitScale(zoom.value))
  const initRotate = 0
  const scale = ref(1)
  const rotate = ref(initRotate)
  const rotateFactor = {
    left: -1,
    right: 1,
  } as const
  const scaleFactor = {
    in: 1,
    out: -1,
  }

  watchEffect(() => (scale.value = initScale.value))

  const scaleDisabled = computed(() => ({
    in: scale.value >= zoom.value[1],
    out: scale.value <= zoom.value[0],
  }))
  const calcTransform = computed(() => ({ transform: `scale(${scale.value}) rotate(${rotate.value}deg)` }))

  const rotateHandle = debounce((direction: RotateType = 'left', rotateStep = 90) => {
    rotate.value = rotate.value + rotateStep * rotateFactor[direction]
  }, debounceTime)

  const scaleHandle = debounce((direction: ScaleType, scaleStep = 0.2) => {
    if (scaleDisabled.value[direction]) {
      return
    }
    scale.value = scale.value + scaleStep * scaleFactor[direction]
  }, debounceTime)

  const resetTransform = () => {
    scale.value = initScale.value
    rotate.value = initRotate
  }

  return {
    calcTransform,
    scaleDisabled,
    rotateHandle,
    scaleHandle,
    resetTransform,
  }
}

function useZoomRange(props: ImageViewerContentProps, config: ImageViewerConfig) {
  return computed(() => props.zoom ?? config.zoom)
}

function useMaskClosable(props: ImageViewerContentProps, config: ImageViewerConfig) {
  return computed(() => props.maskClosable ?? config.maskClosable)
}

function getImageEvent(
  visible: Ref<boolean>,
  {
    setVisible,
    scaleHandle,
    goHandle,
  }: {
    setVisible: (visible: boolean) => void
    scaleHandle: (direction: ScaleType, step?: number) => void
    goHandle: (direction: GoType) => void
  },
) {
  const scroll = (e: WheelEvent | Event) => {
    if (!visible.value) {
      return
    }
    const event = e as WheelEvent & { wheelDelta?: number }
    event.preventDefault()
    const delta = event.wheelDelta ?? -event.detail
    if (delta > 0) {
      scaleHandle('in', 0.2)
    } else {
      scaleHandle('out', 0.2)
    }
  }

  const keyHandle: Record<string, () => void> = {
    ArrowUp: () => scaleHandle('in', 0.2),
    ArrowDown: () => scaleHandle('out', 0.2),
    ArrowLeft: () => goHandle('previous'),
    ArrowRight: () => goHandle('next'),
    Escape: () => setVisible(false),
  }
  const keyDown = (e: KeyboardEvent) => {
    if (!visible.value) {
      return
    }
    e.preventDefault()
    if (e.code in keyHandle) {
      keyHandle[e.code]()
    }
  }

  return {
    onWheelScroll: scroll,
    onKeydown: keyDown,
  }
}

function getInitScale(zoom: number[]) {
  const defaultScale = 1
  if (zoom[0] > defaultScale) {
    return zoom[0]
  }
  if (zoom[1] < defaultScale) {
    return zoom[1]
  }
  return defaultScale
}
