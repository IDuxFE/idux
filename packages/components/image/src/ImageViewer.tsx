/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ImageViewerProps } from './types'
import type { ImageViewerConfig } from '@idux/components/config'
import type { ComputedRef } from 'vue'

import { Transition, computed, defineComponent } from 'vue'

import { CdkPortal } from '@idux/cdk/portal'
import { useControlledProp } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'

import ImageViewerContent from './component/ImageViewerContent'
import { imageViewerProps } from './types'

export default defineComponent({
  name: 'IxImageViewer',
  props: imageViewerProps,
  setup(props) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('imageViewer')
    const [visible] = useControlledProp(props, 'visible', false)
    const mergedPrefixCls = computed(() => `${common.prefixCls}-image-viewer`)
    const target = useTarget(props, config, mergedPrefixCls)

    return () => (
      <CdkPortal target={target.value} load={visible.value}>
        <Transition name={`${common.prefixCls}-zoom-big-fast`} appear>
          {visible.value && <ImageViewerContent mergedPrefixCls={mergedPrefixCls.value} {...props} />}
        </Transition>
      </CdkPortal>
    )
  },
})

function useTarget(props: ImageViewerProps, config: ImageViewerConfig, mergedPrefixCls: ComputedRef<string>) {
  return computed(() => props.target ?? config.target ?? `${mergedPrefixCls.value}-container`)
}
