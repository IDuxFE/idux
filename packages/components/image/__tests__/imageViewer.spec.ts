import type { ImageViewerProps } from '../src/types'
import type { MountingOptions } from '@vue/test-utils'

import { flushPromises, mount } from '@vue/test-utils'

import { renderWork, wait } from '@tests'

import ImageViewer from '../src/ImageViewer'
import ImageViewerContent from '../src/component/ImageViewerContent'

describe('ImageViewer', () => {
  beforeEach(() => {
    const el = document.createElement('div')
    el.className = 'ix-image-viewer-container'
    document.body.appendChild(el)
  })

  afterEach(() => {
    ;(document.querySelector('.ix-image-viewer-container') as HTMLElement).innerHTML = ''
  })

  const ImageViewerMount = (options: MountingOptions<Partial<ImageViewerProps>> = {}) => {
    const { props = {}, ...rest } = options
    const images = ['/icons/logo.svg']
    const activeIndex = 0
    return mount(ImageViewer, {
      ...({
        ...rest,
        props: { images, activeIndex, ...props },
      } as MountingOptions<ImageViewerProps>),
    })
  }

  renderWork<ImageViewerProps>(ImageViewer, { props: { images: ['/icons/logo.svg'] } })

  test('v-model:visible work', async () => {
    const onUpdateVisible = jest.fn()
    const wrapper = ImageViewerMount({ props: { visible: false, 'onUpdate:visible': onUpdateVisible } })
    await flushPromises()

    expect(wrapper.findComponent(ImageViewerContent).exists()).toBe(false)

    await wrapper.setProps({ visible: true })

    expect(wrapper.findComponent(ImageViewerContent).exists()).toBe(true)

    await wrapper.findComponent(ImageViewerContent).find('img').trigger('click')

    expect(onUpdateVisible).toBeCalledWith(false)

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-close').trigger('click')

    expect(onUpdateVisible).toBeCalledWith(false)

    await wrapper.findComponent(ImageViewerContent).trigger('keydown.esc')

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('v-model:activeIndex work', async () => {
    const onUpdateActiveIndex = jest.fn()
    const images = ['/1.png', '/2.png', '/3.png']
    const wrapper = ImageViewerMount({
      props: { images, visible: true, activeIndex: 0, 'onUpdate:activeIndex': onUpdateActiveIndex },
    })
    await flushPromises()

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('src')).toBe(images[0])

    await wrapper.setProps({ activeIndex: 1 })

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('src')).toBe(images[1])

    await wrapper.findComponent(ImageViewerContent).trigger('keydown', { code: 'ArrowRight' })
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(2)

    await wrapper.findComponent(ImageViewerContent).trigger('keydown', { code: 'ArrowLeft' })
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(0)

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-left').trigger('click')
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(0)

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-right').trigger('click')
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(2)
  })

  test('images work', async () => {
    const imagesOld = ['/1.png']
    const wrapper = ImageViewerMount({
      props: { images: imagesOld, visible: true, activeIndex: 0 },
    })
    await flushPromises()

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('src')).toBe(imagesOld[0])

    const imageNew = ['/2.png']
    await wrapper.setProps({ images: imageNew })

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('src')).toBe(imageNew[0])
  })

  test('zoom work', async () => {
    const wrapper = ImageViewerMount({ props: { visible: true } })
    await flushPromises()

    await wrapper.findComponent(ImageViewerContent).trigger('mousewheel', { wheelDelta: 10 })
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1.2)')

    await wrapper.findComponent(ImageViewerContent).trigger('mousewheel', { wheelDelta: -10 })
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1)')

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-zoom-in').trigger('click')
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1.2)')

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-zoom-out').trigger('click')
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1)')

    await wrapper.setProps({ zoom: [0.8, 0.9] })

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(0.9)')

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-zoom-in').trigger('click')
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(0.9)')

    await wrapper.setProps({ zoom: [1.1, 1.2] })

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1.1)')

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-zoom-out').trigger('click')
    await wait(10) // debounce

    expect(wrapper.findComponent(ImageViewerContent).find('img').attributes('style')).toMatch('scale(1.1)')
  })

  test('loop work', async () => {
    const images = ['/1.png', '/2.png', '/3.png']
    const onUpdateActiveIndex = jest.fn()
    const wrapper = ImageViewerMount({
      props: { images, visible: true, activeIndex: 0, 'onUpdate:activeIndex': onUpdateActiveIndex },
    })
    await flushPromises()

    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-left').trigger('click')
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(2)

    await wrapper.setProps({ activeIndex: 2 })
    await wrapper.findComponent(ImageViewerContent).find('.ix-icon-right').trigger('click')
    await wait(10) // debounce

    expect(onUpdateActiveIndex).toBeCalledWith(0)

    await wrapper.setProps({ loop: false })

    expect(wrapper.findComponent(ImageViewerContent).find('.ix-icon-right').classes().toString()).toMatch('disabled')

    await wrapper.setProps({ activeIndex: 0 })

    expect(wrapper.findComponent(ImageViewerContent).find('.ix-icon-left').classes().toString()).toMatch('disabled')
  })

  test('maskClosable work', async () => {
    const onUpdateVisible = jest.fn()
    const wrapper = ImageViewerMount({
      props: { visible: true, 'onUpdate:visible': onUpdateVisible },
    })
    await wrapper.findComponent(ImageViewerContent).find('img').trigger('click')

    expect(onUpdateVisible).toBeCalledWith(false)
    onUpdateVisible.mockRestore()

    await wrapper.setProps({ maskClosable: false })
    await wrapper.findComponent(ImageViewerContent).find('.ix-image-viewer-preview').trigger('click')

    expect(onUpdateVisible).not.toBeCalled()
  })

  test('target work', async () => {
    const wrapper = ImageViewerMount({ props: { visible: true } })
    await flushPromises()

    expect((document.querySelector('.ix-image-viewer-container .ix-image-viewer') as HTMLElement).innerHTML).not.toBe(
      '',
    )

    await wrapper.setProps({ target: 'image-viewer-container' })

    expect((document.querySelector('.image-viewer-container .ix-image-viewer') as HTMLElement).innerHTML).not.toBe('')
  })
})
