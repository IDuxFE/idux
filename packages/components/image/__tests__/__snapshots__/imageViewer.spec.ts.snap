// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageViewer render work 1`] = `"<!---->"`;
